### PR TITLE
Merge LoCo with Zero++

### PR DESCRIPTION
**Pull request by XingyuXie**
#### Overview
This PR introduces the integration of the **LoCo** method, as outlined in [this paper](https://arxiv.org/abs/2407.04480), into the ZeRO++ framework of DeepSpeed. The key enhancement involves applying error feedback compensation to 4-bit gradients before communication. This approach ***improves pre-training loss outcomes without additional time overhead***, though it requires extra GPU memory. The extent of this memory increase depends on model size and training configuration.

#### Experimental Results
We conducted pre-training experiments using the Llama2 architecture, adjusting the number of layers and hidden size. The experiments included:
- **A smaller-scale model with 0.8B parameters trained on 30B tokens**.
- **A larger-scale model with 8B parameters trained on 5B tokens**.

The training data was sampled from **Redpajama-V2**.
<p align="center">
  
  
</p>

**Findings**:
- **Smaller Models (0.8B parameters)**: Significant gains were observed when applying the LoCo method.
- **Larger Models (8B parameters)**: The gains were present but less pronounced. This could be due to:
  1. Relatively smaller data volume.
  2. Lower pre-training loss for larger models, making significant improvements harder to achieve.

However, even a smaller pre-training loss gap in larger models can translate to meaningful gains in downstream tasks.

#### Example Script
For reference, the [run.sh](https://github.com/user-attachments/files/17679552/zeroplus-7b3.zip) script used for the 8B parameter, 5B tokens experiment is attached. The experiment was conducted using the **DeepSpeed-Megatron** platform.

#### Acknowledgments
Special thanks to cc @GuanhuaWang for ongoing communication and guidance throughout this work.

---

We appreciate your consideration of this PR and welcome any feedback or questions!

### CODE REVIEW

**Comment by XingyuXie** [comment]
> @XingyuXie please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

File: deepspeed/runtime/zero/config.py [hunk 1]
```diff
@@ -301,6 +302,10 @@ class DeepSpeedZeroConfig(DeepSpeedConfigModel):
     Boolean indicating whether to use quantized zero gradients
     for efficient all_2_all_reduce comm
     """
+    zeropp_loco_param: Optional[Dict[str, Any]] = None
+    """
+    loco-zero++  params
```
Comment (GuanhuaWang) [inline deepspeed/runtime/zero/config.py:307]: maybe having a brief introduction of two params loco needed and what is prefered default values

**Review by GuanhuaWang — COMMENTED** [review]
overall looks good to me. Just left a few comments

File: deepspeed/runtime/zero/stage3.py [hunk 2]
```diff
@@ -1379,7 +1382,11 @@ def __avg_scatter_grads(self, params_to_reduce: List[Parameter]) -> List[Tensor]
         global_world_size = dist.get_world_size()
         num_nodes = global_world_size // local_world_size
         if self.all2all_process_group is not None and num_nodes > 1:
-            grad_partitions_for_rank = all_to_all_quant_reduce(full_grads_for_rank, self.all2all_process_group)
+            grad_partitions_for_rank = (
+                all_to_all_loco_quant_reduce(params_to_reduce, self.all2all_process_group, self.zeropp_loco_param)
```
[the remark this answers is not in the document]
Comment (XingyuXie) [reply deepspeed/runtime/zero/stage3.py:1386]: 1. This is because we want to use `param.xxx` within the `all_to_all_loco_quant_reduce` function to store, access, and modify the error feedback buffer corresponding to each parameter. For example, in the function, we use `p.inter_ef_buf[0]`. Since `full_grads_for_rank` is a list of tensors, assigning or maintaining attributes directly to it is not convenient.

2. Another reason is to minimize modifications to `stage3.py`. By restricting changes to our own functions, we reduce the risk of introducing unintended bugs.

File: csrc/includes/quantization_utils.h [hunk 3]
```diff
@@ -145,112 +146,85 @@ class Params<Type::Asymmetric, numBits> {
...
 
-template <>
-class GroupStats<Type::Asymmetric> {
+// Asymmetric Quantization
+template <typename DataType>
+class GroupStats<Type::Asymmetric, DataType> {
 public:
-    __half2 cur_max;
-    __half2 cur_min;
+    DataType cur_max;
```
[the remark this answers is not in the document]
Comment (XingyuXie) [reply csrc/includes/quantization_utils.h:207]: 1. In the `loco_swizzled_quant_kernel` and `loco_dequant_reduce` functions, we use the `__half` type for `dequantize::chunk` instead of `__half2`. This is because the gradients have already been quantized to 4-bit precision, and we hypothesize that using `__half` precision for dequantization is sufficient, without needing `__half2`. The goal is to reduce computation and memory usage further, though we haven't profiled to confirm specific gains.

2. Since `__half` is used for `dequantize::chunk`, we needed a version of `quantize::GroupStats` that supports the `__half` type, using like `quantize::GroupStats<quantType, __half> stats`. By defining templates, we ensure support for the `__half` type `GroupStats` without affecting any existing usage of the original `quantize::GroupStats` methods.

File: deepspeed/runtime/zero/config.py (hunk 1 above)
Comment (XingyuXie) [reply deepspeed/runtime/zero/config.py]: Thanks for the advice, we already add the basic description here in the update.

File: csrc/includes/quantization_utils.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (hwchen2017) [reply csrc/includes/quantization_utils.h:203]: `__half2` is a vector of 2 `__half`, not a higher precision of `__half`. Using `__half` will not reduce memory usage because they are the same thing.

File: csrc/quantization/swizzled_quantize.cu [hunk 4]
```diff
@@ -194,3 +195,222 @@ void launch_swizzled_quant(int8_t* q_data,
...
+            // Dequantize the quantized output to compute the dequantized value
+            __half dequant_buffer[quantize::h_per_load];
+            dequantize::chunk<__half, numBits, quantType>(dequant_buffer, local_output, de_params);
+
+// Compute new error: sum - dequant_buffer
+#pragma unroll
+            for (int j = 0; j < quantize::h_per_load; j++) {
+                __half new_error = __hsub(iteration_buffer[j], dequant_buffer[j]);
+                iter_err_buffer[j] = __hmul(iter_err_buffer[j], __float2half(err_beta)) +
+                                     __hmul(__float2half(1.0f - err_beta), new_error);
```
Comment (hwchen2017) [inline csrc/quantization/swizzled_quantize.cu:300]: I guess the difference between `iteration_buffer` and `dequant_buffer` is very small. Will it be beneficial to convert them to `float`, calculate in `float` type, and then convert it back to `__half`?

File: csrc/includes/quantization_utils.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (XingyuXie) [reply csrc/includes/quantization_utils.h:203]: @hwchen2017 U mean we shall change back the to `__half2 ` for the kernel `loco_dequant_reduce` and `loco_swizzled_quant_kernel`?

File: csrc/quantization/swizzled_quantize.cu (hunk 4 above)
Comment (XingyuXie) [reply csrc/quantization/swizzled_quantize.cu:300]: @hwchen2017 Will it cost too much? If not, I can change it and try it. But I do not have the debug env right now.

File: csrc/includes/quantization_utils.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (hwchen2017) [reply csrc/includes/quantization_utils.h:203]: There are no benefits to using `__half` instead of `__half2`. It's fine to leave it as is if everything is working correctly. If you don't plan to change it to `__half2`, please restore the deleted comments, and add a link to your paper in the appropriate places.

File: csrc/quantization/swizzled_quantize.cu (hunk 4 above)
Comment (XingyuXie) [reply csrc/quantization/swizzled_quantize.cu]: Thanks for the advice,  
 - I already use the float precision to update the error feedback buffer
 - I change back to __half2 format. Really, thanks for correcting my previous misunderstanding. 
 
I tested the new code on several toy test examples and used it to continually train an 8B Llama2 model. The results seem okay. Please check the code, and I am willing to refine it further according to your comments. 

Thanks again.

File: csrc/quantization/swizzled_quantize.cu [hunk 5]
```diff
@@ -194,3 +195,236 @@ void launch_swizzled_quant(int8_t* q_data,
...
+    const __half* uncompressed_data_base = uncompressed_data + base_offset_orig;
+
+    const int block_rank_swizzled =
+        blockIdx_x + blockIdx_y * gridDim.x + blockIdx_z_swizzled * gridDim.x * gridDim.y;
+    const int block_offset_swizzled = block_rank_swizzled * elems_per_group;
+    const int base_offset_swizzled = block_offset_swizzled + elem_offset;
+    __half* error_feedback_base = error_feedback + base_offset_swizzled;
+
+    __half2 local_buffer[totalChunks * quantize::h_per_load];
+    __half2 err_buffer[totalChunks * quantize::h_per_load];
```
Comment (hwchen2017) [inline csrc/quantization/swizzled_quantize.cu:239]: You use both `quantize::h_per_load` and `quantize::h2_per_load` in this kernel code. Can you double check to make sure that both of them were used correctly? This one doesn't match the code line 245 and 246.

**Comment by XingyuXie** [comment]
As required by cc @GuanhuaWang , we added the `unit_test` code to verify the logic of cuda kernels used in LoCo-Zero++.

**Comment by GuanhuaWang** [comment]
Thx @XingyuXie for the pr updates on unit-test. Overall, it Looks good to me. 

cc @tjruwase @hwchen2017

**Review by hwchen2017 — APPROVED** [review]
